### ISSUE TITLE
Enabling XLA in tensorflow 2.16 causes memory leaks

### ISSUE THREAD

**Issue by Di-Is** [tensorflow/tensorflow#64170]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

tensorflow 2.16.1
(tensorflow[and-cuda] 2.16.1)

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.12

### CUDA/cuDNN version

12.3/8.9.7

### GPU model and memory

RTX3060Ti、RTX3060

### Current behavior?

Executing the tf.keras.Model.fit method with XLA enabled will cause a memory leak.
Note that XLA seems to be enabled by default since Tensorflow 2.16.1.
Setting tf.keras.Model.jit_compile to False disables XLA and eliminates the memory leak.

I think that updating to tensorflow 2.16 will cause memory leaks in almost all existing programs.
I think you need to fix this problem as soon as possible or alert people about it in documents such as Release Note or README.

### Standalone code to reproduce the issue

```python
import numpy as np
import tensorflow as tf
import gc


def call(epochs: int):
    x = np.arange(-1, 1, 0.01)
    y = 0.8 * x + 0.2
    model = tf.keras.Sequential([tf.keras.layers.Dense(1, activation=None)])
    model.compile("sgd", "mse")
    # model.jit_compile = False
    model.build(input_shape=(0,1))
    model.fit(x, y, epochs=epochs)
    del model

epoch_num = 10000
for i in range(epoch_num):
    call(1)
    tf.keras.backend.clear_session()
    gc.collect()
```

**Comment by NBCBM** [comment]
To address the memory leak issue you're encountering with TensorFlow 2.16.1 and XLA enabled, you can disable XLA by setting tf.keras.Model.jit_compile to False. Here's how you can modify your code to implement this workaround: import numpy as np
import tensorflow as tf
import gc

def call(epochs: int):
    x = np.arange(-1, 1, 0.01)
    y = 0.8 * x + 0.2
    model = tf.keras.Sequential([tf.keras.layers.Dense(1, activation=None)])
    model.compile("sgd", "mse")
    
    # Disable XLA to prevent memory leak
    model.jit_compile = False
    
    model.build(input_shape=(0,1))
    model.fit(x, y, epochs=epochs)
    del model

epoch_num = 10000
for i in range(epoch_num):
    call(1)
    tf.keras.backend.clear_session()
    gc.collect()

**Comment by Di-Is** [comment]
@NBCBM 
Please do not post LLM generated text.

**Comment by Venkat6871** [comment]
Hi **@Di-Is** ,
I tried to run your code on Colab using TF v2.16 and i am not facing any issue. Please find the [gist](https://colab.sandbox.google.com/gist/Venkat6871/a413588819155a7f565bf4f09ac4b5db/64170_2-16-v.ipynb) here for reference.

Thank you!

**Comment by Di-Is** [comment]
@Venkat6871

Thank you for your reply!
I am using tensorflow compiled for Nvidia GPU(tensorflow[and-cuda] 2.16.1).
It seems there is also a difference in the Python version you confirmed.

**Comment by Di-Is** [comment]
I observed memory leaks in [Google Colab with T4 GPU](https://colab.research.google.com/gist/Di-Is/bcc17f37f8c88639317da9b503d25beb/tf216_memory_leak.ipynb).

**Comment by juanma9613** [comment]
@Venkat6871, do you have any progress on this issue.

I'm also affected because of this issue, I tried will all tf versions from 2.11 up to 2.16 and it seems like this happens since tf 2.12. It seems this was not happening for tf 2.11

**Comment by sgkouzias** [comment]
I face the exact problem. Using Ubuntu 20.04, NVIDIA RTX3060 & python=3.11

@Venkat6871, do you have any progress on this issue?

**Comment by AshwinAmbal** [comment]
This document from NVIDIA on tweaking environment variables for XLA memory could help. We're yet to test this out though and will try to keep this thread updated on any findings: 
https://docs.nvidia.com/deeplearning/frameworks/tensorflow-user-guide/index.html#xla-best-practices

Btw, the formation of clusters / Operation fusing / compilation using XLA increases the memory for each different shape of input that the framework comes across. If we keep the shape constant, we've found that after X amount of time, the memory growth stabilizes even with TF 2.15.

**Comment by BobbyWilt** [comment]
I'm also encountering the exact same problem using Ubuntu 20.04 NVIDIA GTX1070 and python 3.10.  Would be great to get this fixed since the memory leakage can get pretty astronomical if left unchecked.  I've had it grow from up to 12gB until it crashed my kernel.

**Comment by SuryanarayanaY** [comment]
Hi @Di-Is ,

I have replicated the reported memory leak with XLA. Attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/a63cfe84f71920b5364878c338f22990/64170_tf216_memory_leak.ipynb) for reference.

**Comment by BobbyWilt** [comment]
Any update on this issue?

**Comment by scarlehoff** [comment]
We are also being hit by this issue https://github.com/NNPDF/nnpdf/pull/2127#issuecomment-2240930598

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by Di-Is** [comment]
@Venkat6871 

My reply was delayed, but I confirmed using the T4 Runtime in Google Colab, and the memory leak persists even in the latest version.

Could you please reopen the issue?

**Comment by GodlyDonuts** [comment]
I reproduced the growth on TensorFlow 2.21.0 / Keras 3.15.1, Python 3.11, with a V100 GPU. Using the issue’s repeated-new-model pattern (one Dense layer, one training epoch, `clear_session()` and `gc.collect()` after each model), process RSS at iterations 5→40 grew from 854→934 MiB with `jit_compile=True`, versus 845→860 MiB with `jit_compile=False`. Repeating 40 fits on *one* XLA-compiled model instead stayed approximately flat after warm-up (841 MiB at iteration 5, 834 MiB at 40). TensorFlow GPU allocator current usage was near zero after each cleanup, so the observed persistent growth is host-side. `tensorflow/compiler/jit/device_compilation_cache.h` explicitly notes that its compiled-executable cache currently has no eviction and can grow without bound; each new model can introduce a distinct compilation signature. This is a plausible retention path, though the RSS measurement alone does not prove every retained byte comes from that cache. A fix needs to address cache ownership/eviction for obsolete model functions without invalidating in-flight compilations.
